src/dune_rules/lib.ml

```diff
@@ -778,15 +778,18 @@ end = struct
     let open R.O in
     fun ts impls ->
       let rec loop t =
+        let _direct = snd t in
+        let t = fst t in
         let t = Option.value ~default:t (Map.find impls t) in
         let* res, visited = R.get in
         if Id.Set.mem visited t.unique_id
         then R.return ()
         else
           let* () = R.set (res, Id.Set.add visited t.unique_id) in
           let* deps = R.lift (Memo.return t.requires) in
+          let deps = List.map ~f:(fun lib -> (lib, true)) deps in
           let* () = many deps in
-          R.modify (fun (res, visited) -> t :: res, visited)
+          R.modify (fun (res, visited) -> (t , true) :: res, visited)
```

Comment:
This change looks a bit fishy, the boolean is added to `deps` and then removed from it just above (`t = fst t`); this can probably be simplified. But does not need to be done right away; first get it to compile.